compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,602 @@
+  buildDefUseMaps(symSet, symExprs, defMap, useMap);
+
+  //data structures for deferred actual handling
+  std::set<CallExpr*> deferredFns;
+  ActualUseDefCastMap actualUseDefMap;
+
+  bool cachedGlobalManip = globalManipFuncCache.count(fn);
+
+  forv_Vec(Symbol, sym, symSet) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar;
+    Type* castTo = NULL;
+
+    //if we don't already have it cached,
+    //check for global symbols in function body
+    if(!cachedGlobalManip) {
+      //I am not sure if there should be !isConstant and !isParam as well
+      //I think there can be a const but stateful global that is manipulated by
+      //the function so, play safe
+      if(sym && !sym->isImmediate() && isGlobal(sym)){
+        globalManipFuncCache[fn] = true;
+        cachedGlobalManip = true;
+      }
+    }
+
+    if(isDenormalizable(sym, defMap, useMap, &use, &def, &castTo)) {
+      usePar = use->parentExpr;
+      defPar = def->parentExpr;
+
+      //defer if the symbol used as actual
+      if(CallExpr* useCe = toCallExpr(usePar)){
+        if(!useCe->isPrimitive()) {
+          deferredFns.insert(useCe);
+          std::pair<Expr*, Type*> defCastBundle(def, castTo);
+          actualUseDefMap.insert(std::pair<SymExpr*, std::pair<Expr*, Type*> >
+              (use, defCastBundle));
+          continue;
+        }
+      }
+
+      //denormalize if the def is safe to move and there is no unsafe function
+      //between use and def
+      if(isExprSafeForReorder(def)) {
+        if(!possibleDepInBetween(defPar, usePar)) {
+          denormalize(def, use, castTo);
+        }
+      }
+    }
+  } // end loop for symbol
+
+  if(!cachedGlobalManip) {
+    globalManipFuncCache[fn] = false;
+  }
+
+  //handle deferred actuals
+  for(std::set<CallExpr*>::iterator ceIt = deferredFns.begin() ;
+      ceIt != deferredFns.end() ; ceIt++) {
+    denormalizeActuals(*ceIt, defMap, useMap, actualUseDefMap);
+  }
+
+  freeDefUseMaps(defMap, useMap);
+}
+
+void denormalizeActuals(CallExpr* ce,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap,
+    ActualUseDefCastMap actualUseDefMap) {
+
+  INT_ASSERT(!ce->isPrimitive());
+  for_alist_backward(actual, ce->argList) {
+    if(SymExpr* argSym = toSymExpr(actual)) { //else it's already denormd
+      if(! (argSym->var->isConstant() || argSym->var->isParameter())) {
+        if(actualUseDefMap.count(argSym) > 0) {
+          std::pair<Expr*,Type*> tmpTuple = actualUseDefMap[argSym];
+          SymExpr* use = argSym;
+          Expr* def = tmpTuple.first;
+          Type* castTo = tmpTuple.second;
+          Expr* usePar = use->parentExpr;
+          Expr* defPar = def->parentExpr;
+          if(CallExpr* ceTmp = toCallExpr(defPar)) {
+            if(!isRecord(ceTmp->get(1)->typeInfo())) { //to preserve pass-by-value
+              if(isExprSafeForReorder(def)) {
+                if(!possibleDepInBetween(defPar, usePar)) {
```

Comment:
This psosibleDepInBetween business is written as a for loop, which means that the analysis part of this algorithm is O(n*2) where n is the number of expressions in a function.

One way to change it to be O(n) is to create a maximum distance between the def and the use (e.g. stop if more than 10 expressions are in-between).

Another way would be to create a count of the number of operations that modify state at each position in the function/block.

E.g. if we have:

    var a = someSideEffectingCall();
    var b = 1;
    var c = 2;
    someSideEffectingCall();

First, record whether or not each statement would inhibit reordering:

    BAD  var a = someSideEffectingCall();
     OK   var b = 1;
     OK   var c = 2;
    BAD  someSideEffectingCall();

Then, assign the value 0 to OK (doesn't inhibit reordering) and 1 to BAD (does inhibit reordering):


     1  var a = someSideEffectingCall();
     0  var b = 1;
     0  var c = 2;
     1  someSideEffectingCall();

Now, compute from these values to a running total:

     1  1 var a = someSideEffectingCall();
     1  0 var b = 1;
     1  0 var c = 2;
     2  1 someSideEffectingCall();

Now, when asking whether or not two statements can be reordered, we can compare this running total to find out if there were any statements in-between that had a side effect. If the running total is equal, there were no such statements. (Of course you have to better define "between" and handle the boundaries correctly).